File: tests/cuda/cpp/test_signal_windows.cpp
// tests/cuda/test_signal_windows.cpp
#include "insight/insight.h"
#include <cmath>
#include <gtest/gtest.h>
#include <vector>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace ins;

namespace {

class SignalWindowsTestGPU : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    ins::init({"cpu", "iluvatar"});
    try {
      set_device(GPUPlace(0));
      // Iluvatar: signal API uses F64
      GTEST_SKIP() << "Iluvatar: signal API uses F64 (hardware limit)";
    } catch (...) {
      GTEST_SKIP() << "GPU not available";
    }
  }
};

// ========== boxcar ==========

TEST_F(SignalWindowsTestGPU, BoxcarBasic) {
  Array w = signal::boxcar(5);
  Array w_cpu = w.to(CPUPlace());
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_DOUBLE_EQ(w_cpu.data<double>()[i], 1.0);
  }
}

TEST_F(SignalWindowsTestGPU, BoxcarSize1) {
  Array w = signal::boxcar(1);
  Array w_cpu = w.to(CPUPlace());
  EXPECT_EQ(w_cpu.numel(), 1);
  EXPECT_DOUBLE_EQ(w_cpu.data<double>()[0], 1.0);
}

TEST_F(SignalWindowsTestGPU, BoxcarPeriodic) {
  Array w = signal::boxcar(5, false);
  Array w_cpu = w.to(CPUPlace());
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_DOUBLE_EQ(w_cpu.data<double>()[i], 1.0);
  }
}

// ========== triang ==========

TEST_F(SignalWindowsTestGPU, TriangOdd) {
  Array w = signal::triang(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 1.0 / 3.0, 1e-10);
  EXPECT_NEAR(d[1], 2.0 / 3.0, 1e-10);
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[3], 2.0 / 3.0, 1e-10);
  EXPECT_NEAR(d[4], 1.0 / 3.0, 1e-10);
}

TEST_F(SignalWindowsTestGPU, TriangEven) {
  Array w = signal::triang(4);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.25, 1e-10);
  EXPECT_NEAR(d[1], 0.75, 1e-10);
  EXPECT_NEAR(d[2], 0.75, 1e-10);
  EXPECT_NEAR(d[3], 0.25, 1e-10);
}

// ========== parzen ==========

TEST_F(SignalWindowsTestGPU, ParzenBasic) {
  Array w = signal::parzen(8);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.00390625, 1e-6);
  EXPECT_NEAR(d[1], 0.10546875, 1e-6);
  EXPECT_NEAR(d[2], 0.47265625, 1e-6);
  EXPECT_NEAR(d[3], 0.91796875, 1e-6);
  EXPECT_NEAR(d[4], d[3], 1e-10);
  EXPECT_NEAR(d[5], d[2], 1e-10);
  EXPECT_NEAR(d[6], d[1], 1e-10);
  EXPECT_NEAR(d[7], d[0], 1e-10);
}

// ========== bohman ==========

TEST_F(SignalWindowsTestGPU, BohmanEndpoints) {
  Array w = signal::bohman(11);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.0, 1e-10);
  EXPECT_NEAR(d[10], 0.0, 1e-10);
  EXPECT_NEAR(d[5], 1.0, 1e-6);
}

// ========== bartlett ==========

TEST_F(SignalWindowsTestGPU, BartlettBasic) {
  Array w = signal::bartlett(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.0, 1e-10);
  EXPECT_NEAR(d[1], 0.5, 1e-10);
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[3], 0.5, 1e-10);
  EXPECT_NEAR(d[4], 0.0, 1e-10);
}

// ========== cosine ==========

TEST_F(SignalWindowsTestGPU, CosineBasic) {
  Array w = signal::cosine(4);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], std::sin(M_PI / 8.0), 1e-10);
  EXPECT_NEAR(d[1], std::sin(3.0 * M_PI / 8.0), 1e-10);
  EXPECT_NEAR(d[2], std::sin(5.0 * M_PI / 8.0), 1e-10);
  EXPECT_NEAR(d[3], std::sin(7.0 * M_PI / 8.0), 1e-10);
}

// ========== blackman ==========

TEST_F(SignalWindowsTestGPU, BlackmanBasic) {
  Array w = signal::blackman(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.0, 1e-10);
  EXPECT_NEAR(d[4], 0.0, 1e-10);
  EXPECT_NEAR(d[2], 1.0, 1e-6);
  EXPECT_NEAR(d[1], d[3], 1e-10);
}

// ========== hann ==========

TEST_F(SignalWindowsTestGPU, HannBasic) {
  Array w = signal::hann(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.0, 1e-10);
  EXPECT_NEAR(d[4], 0.0, 1e-10);
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[1], 0.5, 1e-10);
  EXPECT_NEAR(d[3], 0.5, 1e-10);
}

// ========== hamming ==========

TEST_F(SignalWindowsTestGPU, HammingBasic) {
  Array w = signal::hamming(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.08, 1e-10);
  EXPECT_NEAR(d[4], 0.08, 1e-10);
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[1], d[3], 1e-10);
}

// ========== kaiser ==========

TEST_F(SignalWindowsTestGPU, KaiserBasic) {
  Array w_rect = signal::kaiser(5, 0.0);
  Array w_cpu = w_rect.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], 1.0, 1e-10);
  }
}

TEST_F(SignalWindowsTestGPU, KaiserSymmetry) {
  Array w = signal::kaiser(11, 5.0);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], d[10 - i], 1e-10);
  }
  EXPECT_NEAR(d[5], 1.0, 1e-10);
  EXPECT_GT(d[0], 0.0);
}

// ========== gaussian ==========

TEST_F(SignalWindowsTestGPU, GaussianBasic) {
  Array w = signal::gaussian(5, 1.0);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[0], d[4], 1e-10);
  EXPECT_NEAR(d[1], d[3], 1e-10);
  EXPECT_NEAR(d[0], std::exp(-2.0), 1e-10);
}

// ========== general_gaussian ==========

TEST_F(SignalWindowsTestGPU, GeneralGaussianBasic) {
  Array w = signal::general_gaussian(5, 1.0, 1.0);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[0], std::exp(-2.0), 1e-10);
}

// ========== general_cosine ==========

TEST_F(SignalWindowsTestGPU, GeneralCosineHFT90D) {
  std::vector<double> HFT90D = {1, 1.942604, 1.340318, 0.440811, 0.043097};
  Array w = signal::general_cosine(64, HFT90D, false);
  EXPECT_EQ(w.numel(), 64);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  double max_val = d[0];
  for (int64_t i = 1; i < 64; ++i) {
    if (d[i] > max_val)
      max_val = d[i];
  }
  EXPECT_GT(max_val, 4.0);
  EXPECT_NEAR(d[0], 0.0, 1e-10);
  EXPECT_NEAR(d[63], 0.0, 1e-3);
}

// ========== tukey ==========

TEST_F(SignalWindowsTestGPU, TukeyAlpha0) {
  Array w = signal::tukey(10, 0.0);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  for (int64_t i = 0; i < 10; ++i) {
    EXPECT_NEAR(d[i], 1.0, 1e-10);
  }
}

TEST_F(SignalWindowsTestGPU, TukeyAlpha1) {
  Array w_tukey = signal::tukey(10, 1.0);
  Array w_hann = signal::hann(10);
  Array t_cpu = w_tukey.to(CPUPlace());
  Array h_cpu = w_hann.to(CPUPlace());
  const double *dt = t_cpu.data<double>();
  const double *dh = h_cpu.data<double>();
  for (int64_t i = 0; i < 10; ++i) {
    EXPECT_NEAR(dt[i], dh[i], 1e-10);
  }
}

// ========== barthann ==========

TEST_F(SignalWindowsTestGPU, BarthannBasic) {
  Array w = signal::barthann(11);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[5], 1.0, 0.02);
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], d[10 - i], 1e-10);
  }
}

// ========== exponential ==========

TEST_F(SignalWindowsTestGPU, ExponentialBasic) {
  Array w = signal::exponential(5, -1.0, 1.0);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[2], 1.0, 1e-10);
  EXPECT_NEAR(d[0], d[4], 1e-10);
  EXPECT_NEAR(d[1], d[3], 1e-10);
  EXPECT_NEAR(d[0], std::exp(-2.0), 1e-10);
}

// ========== chebwin ==========

TEST_F(SignalWindowsTestGPU, ChebwinBasic) {
  Array w = signal::chebwin(11, 50.0);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_EQ(w_cpu.numel(), 11);
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], d[10 - i], 1e-6);
  }
  for (int64_t i = 0; i < 11; ++i) {
    EXPECT_GE(d[i], 0.0);
    EXPECT_LE(d[i], 1.0 + 1e-6);
  }
  double max_val = 0;
  for (int64_t i = 0; i < 11; ++i) {
    if (d[i] > max_val)
      max_val = d[i];
  }
  EXPECT_NEAR(max_val, 1.0, 1e-6);
}

// ========== taylor ==========

TEST_F(SignalWindowsTestGPU, TaylorBasic) {
  Array w = signal::taylor(32, 4, -30.0, true);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_EQ(w_cpu.numel(), 32);
  for (int64_t i = 0; i < 16; ++i) {
    EXPECT_NEAR(d[i], d[31 - i], 1e-6);
  }
  for (int64_t i = 0; i < 32; ++i) {
    EXPECT_GE(d[i], 0.0);
    EXPECT_LE(d[i], 1.0 + 1e-6);
  }
  double max_val = 0;
  for (int64_t i = 0; i < 32; ++i) {
    if (d[i] > max_val)
      max_val = d[i];
  }
  EXPECT_NEAR(max_val, 1.0, 1e-6);
}

// ========== get_window ==========

TEST_F(SignalWindowsTestGPU, GetWindowBoxcar) {
  Array w = signal::get_window("boxcar", 8);
  Array w_cpu = w.to(CPUPlace());
  for (int64_t i = 0; i < 8; ++i) {
    EXPECT_DOUBLE_EQ(w_cpu.data<double>()[i], 1.0);
  }
}

TEST_F(SignalWindowsTestGPU, GetWindowHann) {
  Array w = signal::get_window("hann", 5);
  Array ref = signal::hann(5, false);
  Array w_cpu = w.to(CPUPlace());
  Array r_cpu = ref.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  const double *r = r_cpu.data<double>();
  for (int64_t i = 0; i < 5; ++i) {
    EXPECT_NEAR(d[i], r[i], 1e-10);
  }
}

TEST_F(SignalWindowsTestGPU, GetWindowKaiser) {
  Array w = signal::get_window("kaiser", 5.0, 11);
  Array ref = signal::kaiser(11, 5.0, false);
  Array w_cpu = w.to(CPUPlace());
  Array r_cpu = ref.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  const double *r = r_cpu.data<double>();
  for (int64_t i = 0; i < 11; ++i) {
    EXPECT_NEAR(d[i], r[i], 1e-10);
  }
}

// ========== nuttall / blackmanharris / flattop ==========

TEST_F(SignalWindowsTestGPU, NuttallBasic) {
  Array w = signal::nuttall(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 0.0003628, 1e-4);
  EXPECT_NEAR(d[4], 0.0003628, 1e-4);
  EXPECT_NEAR(d[2], 1.0, 1e-6);
}

TEST_F(SignalWindowsTestGPU, BlackmanharrisBasic) {
  Array w = signal::blackmanharris(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_NEAR(d[0], 6.0e-5, 1e-4);
  EXPECT_NEAR(d[4], 6.0e-5, 1e-4);
  EXPECT_NEAR(d[2], 1.0, 1e-6);
}

TEST_F(SignalWindowsTestGPU, FlattopBasic) {
  Array w = signal::flattop(5);
  Array w_cpu = w.to(CPUPlace());
  const double *d = w_cpu.data<double>();
  EXPECT_EQ(w_cpu.numel(), 5);
  EXPECT_NEAR(d[0], d[4], 1e-10);
  EXPECT_NEAR(d[1], d[3], 1e-10);
}

// ========== periodic ==========

TEST_F(SignalWindowsTestGPU, HannPeriodic) {
  Array w_sym = signal::hann(10, true);
  Array w_per = signal::hann(10, false);
  Array s_cpu = w_sym.to(CPUPlace());
  Array p_cpu = w_per.to(CPUPlace());
  EXPECT_EQ(s_cpu.numel(), 10);
  EXPECT_EQ(p_cpu.numel(), 10);
  const double *ds = s_cpu.data<double>();
  const double *dp = p_cpu.data<double>();
  EXPECT_NE(ds[9], dp[9]);
}

} // namespace

File: tests/cuda/lua/test_signal_windows.lua
-- Signal Windows CUDA binding tests.
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_windows.lua

local ok_gpu = pcall(function()
  local native = require("_insight")
  native.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

describe("Signal Windows CUDA Tests", function()
  local ins
  setup(function()
    ins = require("insight")
  end)

  -- ========================================================================
  -- Boxcar (2 tests)
  -- ========================================================================

  it("boxcar_basic", function()
    local w = ins.signal.boxcar(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    for i = 0, 4 do
      assert.near(1.0, ins.item(w, i), 1e-10)
    end
  end)

  it("boxcar_size1", function()
    local w = ins.signal.boxcar(1)
    assert.is_not_nil(w)
    assert.are.equal(1, w.numel)
    assert.near(1.0, ins.item(w, 0), 1e-10)
  end)

  -- ========================================================================
  -- Triang (2 tests)
  -- ========================================================================

  it("triang_odd", function()
    local w = ins.signal.triang(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    -- [1/3, 2/3, 1, 2/3, 1/3]
    assert.near(1.0 / 3.0, ins.item(w, 0), 1e-10)
    assert.near(2.0 / 3.0, ins.item(w, 1), 1e-10)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    assert.near(2.0 / 3.0, ins.item(w, 3), 1e-10)
    assert.near(1.0 / 3.0, ins.item(w, 4), 1e-10)
  end)

  it("triang_even", function()
    local w = ins.signal.triang(4)
    assert.is_not_nil(w)
    assert.are.equal(4, w.numel)
    -- [0.25, 0.75, 0.75, 0.25]
    assert.near(0.25, ins.item(w, 0), 1e-10)
    assert.near(0.75, ins.item(w, 1), 1e-10)
    assert.near(0.75, ins.item(w, 2), 1e-10)
    assert.near(0.25, ins.item(w, 3), 1e-10)
  end)

  -- ========================================================================
  -- Parzen (1 test)
  -- ========================================================================

  it("parzen_basic", function()
    local w = ins.signal.parzen(8)
    assert.is_not_nil(w)
    assert.are.equal(8, w.numel)
    assert.near(0.00390625, ins.item(w, 0), 1e-6)
    assert.near(0.10546875, ins.item(w, 1), 1e-6)
    assert.near(0.47265625, ins.item(w, 2), 1e-6)
    assert.near(0.91796875, ins.item(w, 3), 1e-6)
  end)

  -- ========================================================================
  -- Bohman (1 test)
  -- ========================================================================

  it("bohman_endpoints", function()
    local w = ins.signal.bohman(11)
    assert.is_not_nil(w)
    assert.are.equal(11, w.numel)
    assert.near(0.0, ins.item(w, 0), 1e-10)
    assert.near(0.0, ins.item(w, 10), 1e-10)
    assert.near(1.0, ins.item(w, 5), 1e-6)
  end)

  -- ========================================================================
  -- Bartlett (1 test)
  -- ========================================================================

  it("bartlett_basic", function()
    local w = ins.signal.bartlett(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(0.0, ins.item(w, 0), 1e-10)
    assert.near(0.5, ins.item(w, 1), 1e-10)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    assert.near(0.5, ins.item(w, 3), 1e-10)
    assert.near(0.0, ins.item(w, 4), 1e-10)
  end)

  -- ========================================================================
  -- Cosine (1 test)
  -- ========================================================================

  it("cosine_basic", function()
    local w = ins.signal.cosine(4)
    assert.is_not_nil(w)
    assert.are.equal(4, w.numel)
    -- sin(pi/8), sin(3pi/8), sin(5pi/8), sin(7pi/8)
    assert.near(math.sin(math.pi / 8), ins.item(w, 0), 1e-10)
    assert.near(math.sin(3 * math.pi / 8), ins.item(w, 1), 1e-10)
  end)

  -- ========================================================================
  -- Exponential (1 test)
  -- ========================================================================

  it("exponential_basic", function()
    local w = ins.signal.exponential(5, 1.0, 2.0)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    -- Symmetric
    assert.near(ins.item(w, 0), ins.item(w, 4), 1e-10)
    assert.near(math.exp(-2.0), ins.item(w, 0), 1e-10)
  end)

  -- ========================================================================
  -- Blackman (1 test)
  -- ========================================================================

  it("blackman_basic", function()
    local w = ins.signal.blackman(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(0.0, ins.item(w, 0), 1e-10)
    assert.near(0.0, ins.item(w, 4), 1e-10)
    assert.near(1.0, ins.item(w, 2), 1e-6)
  end)

  -- ========================================================================
  -- Nuttall (1 test)
  -- ========================================================================

  it("nuttall_basic", function()
    local w = ins.signal.nuttall(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(0.0003628, ins.item(w, 0), 1e-4)
    assert.near(1.0, ins.item(w, 2), 1e-6)
  end)

  -- ========================================================================
  -- Blackmanharris (1 test)
  -- ========================================================================

  it("blackmanharris_basic", function()
    local w = ins.signal.blackmanharris(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(6.0e-5, ins.item(w, 0), 1e-4)
    assert.near(1.0, ins.item(w, 2), 1e-6)
  end)

  -- ========================================================================
  -- Flattop (1 test)
  -- ========================================================================

  it("flattop_basic", function()
    local w = ins.signal.flattop(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(ins.item(w, 0), ins.item(w, 4), 1e-10)
    assert.near(ins.item(w, 1), ins.item(w, 3), 1e-10)
  end)

  -- ========================================================================
  -- Hann (2 tests)
  -- ========================================================================

  it("hann_basic", function()
    local w = ins.signal.hann(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(0.0, ins.item(w, 0), 1e-10)
    assert.near(0.0, ins.item(w, 4), 1e-10)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    assert.near(0.5, ins.item(w, 1), 1e-10)
    assert.near(0.5, ins.item(w, 3), 1e-10)
  end)

  it("hann_size16", function()
    local w = ins.signal.hann(16)
    assert.is_not_nil(w)
    assert.are.equal(16, w.numel)
  end)

  -- ========================================================================
  -- Hamming (1 test)
  -- ========================================================================

  it("hamming_basic", function()
    local w = ins.signal.hamming(5)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(0.08, ins.item(w, 0), 1e-10)
    assert.near(0.08, ins.item(w, 4), 1e-10)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    assert.near(ins.item(w, 1), ins.item(w, 3), 1e-10)
  end)

  -- ========================================================================
  -- Tukey (2 tests)
  -- ========================================================================

  it("tukey_alpha0", function()
    local w = ins.signal.tukey(10, 0.0)
    assert.is_not_nil(w)
    assert.are.equal(10, w.numel)
    for i = 0, 9 do
      assert.near(1.0, ins.item(w, i), 1e-10)
    end
  end)

  it("tukey_alpha1", function()
    local w_tukey = ins.signal.tukey(10, 1.0)
    local w_hann = ins.signal.hann(10)
    assert.is_not_nil(w_tukey)
    assert.is_not_nil(w_hann)
    for i = 0, 9 do
      assert.near(ins.item(w_hann, i), ins.item(w_tukey, i), 1e-10)
    end
  end)

  -- ========================================================================
  -- Barthann (1 test)
  -- ========================================================================

  it("barthann_basic", function()
    local w = ins.signal.barthann(11)
    assert.is_not_nil(w)
    assert.are.equal(11, w.numel)
    assert.near(1.0, ins.item(w, 5), 0.02)
  end)

  -- ========================================================================
  -- Kaiser (1 test)
  -- ========================================================================

  it("kaiser_beta0", function()
    local w = ins.signal.kaiser(5, 0.0)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    for i = 0, 4 do
      assert.near(1.0, ins.item(w, i), 1e-10)
    end
  end)

  -- ========================================================================
  -- Gaussian (1 test)
  -- ========================================================================

  it("gaussian_basic", function()
    local w = ins.signal.gaussian(5, 1.0)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    assert.near(ins.item(w, 0), ins.item(w, 4), 1e-10)
    assert.near(math.exp(-2.0), ins.item(w, 0), 1e-10)
  end)

  -- ========================================================================
  -- General Gaussian (1 test)
  -- ========================================================================

  it("general_gaussian_basic", function()
    local w = ins.signal.general_gaussian(5, 1.0, 1.0)
    assert.is_not_nil(w)
    assert.are.equal(5, w.numel)
    assert.near(1.0, ins.item(w, 2), 1e-10)
    assert.near(math.exp(-2.0), ins.item(w, 0), 1e-10)
  end)

  -- ========================================================================
  -- Chebwin (1 test)
  -- ========================================================================

  it("chebwin_basic", function()
    local w = ins.signal.chebwin(11, 50.0)
    assert.is_not_nil(w)
    assert.are.equal(11, w.numel)
    -- Symmetric
    for i = 0, 4 do
      assert.near(ins.item(w, i), ins.item(w, 10 - i), 1e-6)
    end
  end)

  -- ========================================================================
  -- Taylor (1 test)
  -- ========================================================================

  it("taylor_basic", function()
    local w = ins.signal.taylor(32, 4, -30, true)
    assert.is_not_nil(w)
    assert.are.equal(32, w.numel)
    -- Symmetric
    for i = 0, 15 do
      assert.near(ins.item(w, i), ins.item(w, 31 - i), 1e-6)
    end
  end)

  -- ========================================================================
  -- Get Window (2 tests)
  -- ========================================================================

  it("get_window_boxcar", function()
    local w = ins.signal.get_window("boxcar", 8)
    assert.is_not_nil(w)
    assert.are.equal(8, w.numel)
    for i = 0, 7 do
      assert.near(1.0, ins.item(w, i), 1e-10)
    end
  end)

  it("get_window_hann", function()
    local w = ins.signal.get_window("hann", 5)
    local ref = ins.signal.hann(5)
    assert.is_not_nil(w)
    assert.is_not_nil(ref)
    for i = 0, 4 do
      assert.near(ins.item(ref, i), ins.item(w, i), 1e-10)
    end
  end)
end)
